### PR TITLE
FL client and main

### PR DESCRIPTION
**Pull request by ZidongLiu**
# Summary
- implement client and main
- share model update in client and server

# screenshot

image

### CODE REVIEW

File: cezo_fl/client.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+class Client(AbstractClient):
+
+    def __init__(
+        self,
+        model,
+        dataloader,
+        device,
+        grad_estimator_params: dict,
+        optimzier_params: dict,
+        criterion,
```
Comment (bichengying) [inline cezo_fl/client.py:19]: Please add type annotation.

File: cezo_fl/client.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            method = grad_estimator_params["method"][4:]
+
+            return RGE(
+                self.model,
+                grad_estimate_method=method,
+                mu=grad_estimator_params["mu"],
+                num_pert=grad_estimator_params["num_pert"],
+                device=self.device,
+            )
+        else:
```
Comment (bichengying) [inline cezo_fl/client.py:45]: prefer early-return style

```python
if method not in [xxx]:
    raise ValueError()
method = grad_estimator_params["method"][4:]
return RGE(
    self.model,
    grad_estimate_method=method,
    mu=grad_estimator_params["mu"],
    num_pert=grad_estimator_params["num_pert"],
    device=self.device,
)
```

File: cezo_fl/client.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    def _initialize_optimizer(self, optimzier_params):
+        if optimzier_params["method"] == "SGD":
+            return SGD(
+                self.model.parameters(),
+                lr=optimzier_params["lr"],
+                momentum=optimzier_params["momentum"],
+                weight_decay=1e-5,
+            )
+        else:
+            raise Exception(f'{optimzier_params["method"]} is not supported')
```
Comment (bichengying) [inline cezo_fl/client.py:57]: same here

File: cezo_fl/client.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                self.model,
+                grad_estimate_method=method,
+                mu=grad_estimator_params["mu"],
+                num_pert=grad_estimator_params["num_pert"],
+                device=self.device,
+            )
+        else:
+            raise Exception(f'{grad_estimator_params["method"]} is not supported')
+
+    def _initialize_optimizer(self, optimzier_params):
```
Comment (bichengying) [inline cezo_fl/client.py:48]: I prefer this initialization put at the main file and just pass the initialized optimizer into this class. Notice we need to make sure the optimizer in clients and server are the same

File: cezo_fl/server.py [hunk 5]
```diff
@@ -169,6 +177,7 @@ def train_one_step(self, iteration: int) -> None:
         self.seed_grad_records.add_records(seeds=seeds, grad=avg_grad_scalar)
 
         # Step 4: client sync-to server's older version.
+        # TODO: Zidong's question, should we just do this at client.pull step?
```
Comment (bichengying) [inline cezo_fl/server.py:180]: I agree that we can put this under the pull step.

File: cezo_fl/client.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        gradient_scalar: Sequence[Sequence[torch.Tensor]],
+    ) -> None:
+        # reset model
+        self.reset_model()
+        # update model to latest version
+        for iteration_seeds, iteration_grad_avgs in zip(seeds_list, gradient_scalar):
+            for local_update_seed, local_update_grad_vector in zip(
+                iteration_seeds, iteration_grad_avgs
+            ):
+                # create gradient
```
Comment (bichengying) [inline cezo_fl/client.py:116]: I think you missed `self.optimizer.zero_grad()`.

Also you can merge this with `update_model_given_seed_and_grad` in server.py into one shared function?

**Review by bichengying — CHANGES_REQUESTED** [review]

File: cezo_fl/shared.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+from typing import Sequence
+from gradient_estimators.random_gradient_estimator import RandomGradientEstimator as RGE
+
+
+def update_model_given_seed_and_grad(
+    model: torch.nn.Module,
+    optimizer: torch.optim.Optimizer,
+    grad_estimator: RGE,
+    seeds_list: Sequence[Sequence[int]],
+    grad_scalar_list: Sequence[Sequence[torch.Tensor]],
```
Comment (bichengying) [inline cezo_fl/shared.py:11]: prefer smaller function, i.e. less functionality for better test and combination.

So it is better to let this execute one iteration and let the caller to decide how many iterations it needs

File: cezo_fl/shared.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import torch
+from typing import Sequence
+from gradient_estimators.random_gradient_estimator import RandomGradientEstimator as RGE
+
+
+def update_model_given_seed_and_grad(
+    model: torch.nn.Module,
```
Comment (bichengying) [inline cezo_fl/shared.py:7]: you don't need to pass the model here

File: cezo_fl/shared.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+            update_grad = (
+                sum(
+                    [
+                        perturb * local_update_grad_vector[j]
+                        for j, perturb in enumerate(perturbations)
+                    ]
+                )
+                / this_num_pert
+            )
```
Comment (bichengying) [inline cezo_fl/shared.py:35]: we can merge this two list-comprehension into one for-loop
```python
update_grad = 0
for j in range(this_num_pert)
    perturb = grad_estimator.generate_perturbation_norm()
    update_grad += perturb * local_update_grad_vector[j] * perturb
grad_estimator.put_grad(update_grad.div_(this_num_pert))
```

IIRC, `div_` should be faster than `div` since it doesn't open a new space

File: cezo_fl/server.py [hunk 10]
```diff
@@ -169,8 +155,9 @@ def train_one_step(self, iteration: int) -> None:
         self.seed_grad_records.add_records(seeds=seeds, grad=avg_grad_scalar)
 
         # Step 4: client sync-to server's older version.
-        for index in sampled_client_index:
-            self.clients[index].reset_model()
+        # Moved this step into client.pull
+        # for index in sampled_client_index:
+        #     self.clients[index].reset_model()
```
Comment (bichengying) [inline cezo_fl/server.py:160]: delete

File: cezo_fl/server.py [hunk 11]
```diff
@@ -122,19 +101,26 @@ def __init__(
         self.seed_grad_records = SeedAndGradientRecords()
         self.client_last_updates = [0 for _ in range(len(self.clients))]
 
-        self.server_model: torch.nn.Module | None = None
-        self.server_criterion: torch.nn.Module | None = None
-        self.optim: torch.optim.Optimizer | None = None
+        self.server_model: Optional[torch.nn.Module] = None
```
Comment (bichengying) [inline cezo_fl/server.py:104]: which python version you used? Now it is preferred to use `x | None` than `Optional[x]`

File: cezo_fl/server.py [hunk 12]
```diff
@@ -179,8 +166,13 @@ def train_one_step(self, iteration: int) -> None:
         )
 
         if self.server_model:
+            # self.train()
```
Comment (bichengying) [inline cezo_fl/server.py:169]: Why comment this out?

File: cezo_fl_main.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+#     return args.lr * current_iterations / overall_iterations
+
+
+if __name__ == "__main__":
+    args = get_params().parse_args()
+
+    clients = []
+
+    for i in range(args.num_clients):
+        device, train_loader, _ = preprocess(args)
```
Comment (bichengying) [inline cezo_fl_main.py:86]: this is wrong. You need to partition the train loader.

File: cezo_fl/client.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+                yield v
+
+    def local_update(self, seeds: Sequence[int]) -> Sequence[torch.Tensor]:
+        """Returns a sequence of gradient scalar tensors for each local update.
+
+        The length of the returned sequence should be the same as the length of seeds.
+        The inner tensor can be a scalar or a vector. The length of vector is the number
+        of perturbations.
+        """
+        ret: Sequence[torch.Tensor] = []
```
Comment (bichengying) [inline cezo_fl/client.py:46]: consider to give a better name

File: cezo_fl/client.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+            batch_inputs, labels = next(self.data_iterator)
+            if self.device != torch.device("cpu"):
+                batch_inputs, labels = batch_inputs.to(self.device), labels.to(
+                    self.device
+                )
+            # generate grads and update model's gradient
+            torch.manual_seed(seed)
+            seed_grads = self.grad_estimator.compute_grad(
+                batch_inputs, labels, self.criterion
+            )
```
Comment (bichengying) [inline cezo_fl/client.py:59]: can you use line limit = 100?

File: gradient_estimators/random_gradient_estimator.py [hunk 16]
```diff
@@ -69,14 +69,16 @@ def put_grad(self, grad: torch.Tensor) -> None:
...
-    def compute_grad(self, batch_inputs, labels, criterion) -> None:
+    def compute_grad(self, batch_inputs, labels, criterion) -> torch.Tensor:
         estimation_method = self.method_func_dict[self.grad_estimate_method]
-        grad = estimation_method(batch_inputs, labels, criterion)
+        grad, dir_grads = estimation_method(batch_inputs, labels, criterion)
 
         self.put_grad(grad)
+        return torch.tensor(dir_grads, device=self.device)
 
     def _forward_method(self, batch_inputs, labels, criterion):
```
Comment (bichengying) [inline gradient_estimators/random_gradient_estimator.py:79]: add return type annotation

File: gradient_estimators/random_gradient_estimator.py [hunk 17]
```diff
@@ -86,11 +88,13 @@ def _forward_method(self, batch_inputs, labels, criterion):
             self.perturb_model(pb_norm, alpha=-self.mu)  # Restore model
 
             dir_grad = (pert_plus_loss - initial_loss) / self.mu
+            dir_grads += [dir_grad]
             grad += dir_grad * pb_norm
-        return grad / self.num_pert
+        return grad / self.num_pert, dir_grads
```
Comment (bichengying) [inline gradient_estimators/random_gradient_estimator.py:93]: isn't `dir_grads` should be concat into one torch tensor?

**Review by bichengying — CHANGES_REQUESTED** [review]

File: cezo_fl/shared.py (hunk 8 above)
Comment (ZidongLiu) [reply cezo_fl/shared.py:7]: good catch

File: cezo_fl/server.py (hunk 11 above)
Comment (ZidongLiu) [reply cezo_fl/server.py:104]: I am using python3.10.14 + black for formatting + flake8 for linting. Do not get a warning for using Optional, Also I do not find a strong opinion today googling which style is preferred. 
For me when reading code, Optional means I can skip this variable by just reading the type, while ' | None' makes me think for a bit before I can realize it's optional.
So I prefer keeping Optional, unless your linter/formatter gives direct warning when using it.

File: cezo_fl/server.py (hunk 12 above)
Comment (ZidongLiu) [reply cezo_fl/server.py:169]: wondering if this is the best place to call it

File: cezo_fl/client.py (hunk 15 above)
Comment (ZidongLiu) [reply cezo_fl/client.py:59]: black's default line limit = 88? Are we sure we want to change it

File: cezo_fl/shared.py (hunk 7 above)
Comment (ZidongLiu) [reply cezo_fl/shared.py]: changed and also updated the server_test

File: cezo_fl/server.py (hunk 12 above)
Comment (ZidongLiu) [reply cezo_fl/server.py]: just uncommentted, let's put self.train here first

File: cezo_fl_main.py (hunk 13 above)
Comment (ZidongLiu) [reply cezo_fl_main.py:86]: Not sure what's the best way to do partition. Remember you and @RogerRogerUSC mentioned something about data distribution, not sure if we need to use that instead

File: cezo_fl/client.py (hunk 15 above)
Comment (ZidongLiu) [reply cezo_fl/client.py]: just changed all settings to 100. But haven't updated each file.

File: cezo_fl_main.py (hunk 13 above)
Comment (RogerRogerUSC) [reply cezo_fl_main.py:86]: There are some related examples in FAST project.

Comment (ZidongLiu) [reply cezo_fl_main.py:86]: Let's leave this for future. 
Checked example from https://github.com/RogerRogerUSC/FAST/blob/main/fashion_main.py, https://github.com/RogerRogerUSC/FAST/blob/main/mnist_main.py, https://github.com/RogerRogerUSC/FAST/blob/main/cifar10_main.py. Find fmnist and mnist are using fedlab while cifar10 implements something on its own.

**Comment by ZidongLiu** [comment]
image

everything still works

File: cezo_fl/shared.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    optimizer.zero_grad()
+    for local_update_seed, local_update_grad_vector in zip(iteration_seeds, iteration_grad_scalar):
+        # create gradient
+        torch.manual_seed(local_update_seed)
+        this_num_pert = local_update_grad_vector.shape[0]
+        update_grad = 0
+        for j in range(this_num_pert):
+            perturb = grad_estimator.generate_perturbation_norm()
+            update_grad += perturb * local_update_grad_vector[j]
+        grad_estimator.put_grad(update_grad.div_(this_num_pert))
```
Comment (bichengying) [inline cezo_fl/shared.py:24]: I just realized this can be further simplified
```python
update_grad = 0
for local_update_grad in local_update_grad_vector:
        perturb = grad_estimator.generate_perturbation_norm()
        update_grad += perturb *local_update_grad
```

File: gradient_estimators/random_gradient_estimator.py [hunk 19]
```diff
@@ -86,11 +88,13 @@ def _forward_method(self, batch_inputs, labels, criterion):
             self.perturb_model(pb_norm, alpha=-self.mu)  # Restore model
 
             dir_grad = (pert_plus_loss - initial_loss) / self.mu
+            dir_grads += [dir_grad]
             grad += dir_grad * pb_norm
-        return grad / self.num_pert
+        return grad / self.num_pert, torch.tensor(dir_grads, device=self.device)
```
Comment (bichengying) [inline gradient_estimators/random_gradient_estimator.py:93]: I don't think this is right. It should be `torch.cat(dir_grads, dim=0)`

Comment (ZidongLiu) [reply gradient_estimators/random_gradient_estimator.py:93]: It's doing what it's supposed to do. See following screenshot for more details.

dir_grads is a list of scalar tensor. Each elements in dir_grads is a scalar tensor, whose shape is [].
Directly use torch.tensor on it creates a 1-d tensor.


image

File: cezo_fl/shared.py (hunk 18 above)
Comment (ZidongLiu) [reply cezo_fl/shared.py:24]: Good suggestion.

I was thinking about something like this to speed up experiment. Let me know if we wa

```suggestion
        this_num_pert = local_update_grad_vector.shape[0]
        update_grad = 0
        for j in range(this_num_pert):
            perturb = iteration_perturbations[j] ## generate perturbation in server and reuse in each clients to save time
            update_grad += perturb * local_update_grad_vector[j]
        grad_estimator.put_grad(update_grad.div_(this_num_pert))
```

File: cezo_fl/server.py (hunk 11 above)
Comment (ZidongLiu) [reply cezo_fl/server.py]: Decided to go with x | None

File: cezo_fl_main.py (hunk 13 above)
Comment (ZidongLiu) [reply cezo_fl_main.py:86]: https://pytorch.org/docs/stable/data.html#torch.utils.data.random_split
https://medium.com/codex/a-comprehensive-tutorial-to-pytorch-distributeddataparallel-1f4b42bb1b51
got some concerns using distributedSampler

Comment (ZidongLiu) [reply cezo_fl_main.py]: resolved in latest commit

**Comment by ZidongLiu** [comment]
image

everything still works

File: preprocess.py [hunk 20]
```diff
@@ -113,14 +109,91 @@ def preprocess(args):
         )
     elif args.dataset == "shakespeare":
         device, kwargs = use_device(args)
-        train_dataset = ShakeSpeare(train=True)
+        trainset = ShakeSpeare(train=True)
```
Comment (bichengying) [inline preprocess.py:112]: LOL, this is [trainset](https://www.google.com/search?sca_esv=14770d8a9caa6d0f&sxsrf=ACQVn08EXj0lLldLlCOCTT2lYM5Ho8Ifjw:1714714960977&q=trainset&uds=AMwkrPs4weTB3nAhxTiOjI0cLaoUD0SWCQBPIAWRLyCpbGKOyTmoaGdJ9DZkPDWjuX5xhbqWsfO062xbows9npjqol77oQisXcKfxIwydZ-agsuJcGIZpXNvLifTxBew2dPseQ4tvRCYhIzDhi6H3L-m19QEhYG2_QITBquv3tQ21NbUAExwyWANWZXJd2uDAPgbi3SvMO3cOCjhY1v8gqMInVz3MlyyA-bI2_5z-Uo_X0B-en2ErKAS22XiTq4_5vh5N23WsOUMIzFH0M990ca9QySm5QIxUGno1Q__L2Jz1bsUGFaFsON3kTnNjnK7I2yLzj5vb_7U&udm=2&prmd=isvnmbtz&sa=X&sqi=2&ved=2ahUKEwimj4-B4_CFAxVuC0QIHRh4ALQQtKgLegQIExAB&biw=1728&bih=992&dpr=2)

Let's keep the standard name

File: preprocess.py [hunk 21]
```diff
@@ -113,14 +109,91 @@ def preprocess(args):
...
+        test_loader = torch.utils.data.DataLoader(
+            testset, batch_size=args.test_batch_size, **kwargs
+        )
+    elif args.dataset == "shakespeare":
+        device, kwargs = use_device(args)
+        trainset = ShakeSpeare(train=True)
+        testset = ShakeSpeare(train=False)
+        test_loader = torch.utils.data.DataLoader(
+            testset, batch_size=args.test_batch_size, shuffle=False, **kwargs
+        )
```
Comment (bichengying) [inline preprocess.py:185]: why not we reuse `preprocess` function here?

Comment (ZidongLiu) [reply preprocess.py:185]: Original preprocess loads 1 test dataset and build 1 dataloader using that dataset. 
However, we need to split that test dataset into num_clients subsets, then build num_clients dataloader. 

Original preprocess return is device: str, train_loader: DataLoader, test_loader: DataLoader
New preprocess_cezo_fl return is device: str, train_loaders: list[DataLoader], test_loader: DataLoader.

return types are different, so I decided to split them into 2 different functions

File: preprocess.py (hunk 20 above)
Comment (ZidongLiu) [reply preprocess.py:112]: Oh, I see we were using 'trainset' in the old preprocess function when args.data != 'shakespeare'. I am going to change all places to train_dataset
